znoyder/get_project_zuul_files.py

```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python3
+
+from argparse import ArgumentParser
+from shutil import rmtree
+import os
+from subprocess import check_call, CalledProcessError
+from sys import exit
+
+URL_BASE_PROJECT = "https://opendev.org/openstack/{project_name}.git"
+
+
+def process_arguments() -> tuple:
+    """ Return the branch, destination folder and project name of the processed arguments"""
+    parser = ArgumentParser()
+    parser.add_argument(
+        '-p', '--projectname',
+        dest='project_name',
+        help="Insert the project name",
+        metavar="PROJECT_NAME",
+        required=True
+    )
+    parser.add_argument(
+        '-b', '--branchname',
+        dest='branch_name',
+        help="Select branch to work",
+        metavar="BRANCH_NAME",
+        required=True
+    )
+    parser.add_argument(
+        '-d', '--destinationfolder',
+        dest='destination_folder',
+        help="Insert the destination folder",
+        metavar="DESTINATION_FOLDER",
+        required=True
+    )
+    arguments = parser.parse_args()
+    return arguments
+
+
+def clone_project(url: str, branch: str, destination_folder: str) -> None:
+    command = ['git', 'clone', '--depth', '1', url, '--branch', branch, destination_folder]
```

Comment:
This will clone entire project. If we want to crawl against 100 projects it's not really efficient. Is there a better way of optimizing that? e.g. using github web interface ? e.g. this is link to raw .zuul.yaml for nova project for master branch:
https://opendev.org/openstack/nova/raw/branch/master/.zuul.yaml

Reply:
I thought about three options:

# Option 1: PyGithub module.

## The module 'PyGithub' can't be used. We can't do the search in OpenDev repository. It works in Github or Github enterprise because it uses the Github APIv3.

For example in the mirror of the Nova project that exists in GitHub.

```
$ -> cat ghub.py 
from github import Github

g = Github()

repo = g.get_repo("openstack/nova")

contents = repo.get_contents("")

for content in contents:
    print(content)
```

```
$ -> python3 ghub.py 
ContentFile(path=".coveragerc")
ContentFile(path=".gitignore")
ContentFile(path=".gitreview")
ContentFile(path=".mailmap")
ContentFile(path=".pre-commit-config.yaml")
ContentFile(path=".stestr.conf")
ContentFile(path=".zuul.yaml")
ContentFile(path="CONTRIBUTING.rst")
ContentFile(path="HACKING.rst")
ContentFile(path="LICENSE")
ContentFile(path="MAINTAINERS")
ContentFile(path="README.rst")
ContentFile(path="api-guide")
ContentFile(path="api-ref")
ContentFile(path="bindep.txt")
ContentFile(path="devstack")
ContentFile(path="doc")
ContentFile(path="etc")
ContentFile(path="gate")
ContentFile(path="lower-constraints.txt")
ContentFile(path="mypy-files.txt")
ContentFile(path="nova")
ContentFile(path="playbooks")
ContentFile(path="releasenotes")
ContentFile(path="requirements.txt")
ContentFile(path="roles")
ContentFile(path="setup.cfg")
ContentFile(path="setup.py")
ContentFile(path="test-requirements.txt")
ContentFile(path="tools")
ContentFile(path="tox.ini")
```

I ruled out this option.

# Option 2: Using `request`.

1. We can do a `GET` request to check:
1.1 if the `.zuul.yaml` file and download the file with a `GET` requests on the raw file.
1.2 Or if `zuul.d` directory exists, parse the HTML, get the .yaml files with regex and download every .yaml file.

It's the worst option.

I ruled out this option.

# Option 3: Using `git` commands.

## Using `git archive`:

It could be a good option:

```
[16-11-21 10:54] {afuscoar@afuscoar} (~)$ -> man -P cat git archive | grep -- '--remote'
                     [--remote=<repo> [--exec=<git-upload-archive>]] <tree-ish>
       --remote=<repo>
           Used with --remote to specify the path to the git-upload-archive on the remote side.
           user’s umask will be used instead. See umask(2) for details. If --remote is used then only the configuration of the remote repository takes effect.
```

But it doesn't work:

```
[16-11-21 10:54] {afuscoar@afuscoar} (...)$ -> git archive --remote=https://opendev.org/openstack/nova.git master
fatal: operation not supported by protocol
```

It works using the `git clone` filtering to avoid big files. 

____________________________________________-

I wrote the script to use the second and third option. 

I'll continue looking for a best way to do it.

Reply:
Option 1: I think it may be not valid. OpenStack projects were hosted in github, but not anymore. They moved to opendev.org. Python github module interacts with github API (api.github.com) which is not provided by opendev.org.

Reply:
Option 3: It's also not valid option as opendev.org remote git does not support archive option.